pootle/apps/virtualfolder/receivers.py

```diff
@@ -31,125 +30,74 @@ def update_vfolder_tree(vf, store):
         vfolder_treeitem.update_all_cache()
 
 
-def add_unit_to_vfolders(unit):
+def add_store_to_vfolders(store):
     """For a given Unit check for membership of any VirtualFolders
     """
-    pootle_path = unit.store.pootle_path
+    pootle_path = store.pootle_path
 
     for vf in VirtualFolder.objects.iterator():
-        unit_added = False
+        store_added = False
         for location in vf.all_locations:
             if not pootle_path.startswith(location):
                 continue
 
             for filename in vf.filter_rules.split(","):
                 if pootle_path == "".join([location, filename]):
-                    vf.units.add(unit)
-                    unit_added = True
+                    vf.stores.add(store)
+                    store_added = True
                     break
 
-            if unit_added:
+            if store_added:
                 break
 
-        if unit_added:
-            update_vfolder_tree(vf, unit.store)
+        if store_added:
+            update_vfolder_tree(vf, store)
 
 
-@receiver(pre_save, sender=VirtualFolder)
-def vfolder_unit_priority_presave_handler(**kwargs):
+@receiver(post_save, sender=VirtualFolder)
+def vfolder_save_handler(sender, instance, created, **kwargs):
     """Remove Units from VirtualFolder when vfolder changes
-
     - Check the original VirtualFolder object's locations
     - Check the new VirtualFolders object's locations
     - Remove any units from the VirtualFolder that are only in original's
       locations and reset Unit priority
     - Update VirtualFolderTree for any Stores that may have been affected
     """
-    instance = kwargs["instance"]
-    if instance.id is None:
-        return
-
-    original = VirtualFolder.objects.get(pk=instance.pk)
-
-    original_locations = set()
-    new_locations = set()
-
-    for location in original.all_locations:
-        for filename in original.filter_rules.split(","):
-            original_locations.add("".join([location, filename]))
-
+    locations = set()
     for location in instance.all_locations:
         for filename in instance.filter_rules.split(","):
-            new_locations.add("".join([location, filename]))
-
-    removed_locations = original_locations - new_locations
-
-    stores_affected = set()
-    for location in removed_locations:
-        # reindex these units without this vfolder
-        removed_units = (
-            Unit.objects.filter(store__pootle_path__startswith=location,
-                                vfolders=original))
-        for unit in removed_units.iterator():
-            unit.vfolders.remove(original)
-            stores_affected.add(unit.store)
-
-    for store in stores_affected:
-        update_vfolder_tree(original, store)
+            locations.add("".join([location, filename]))
```

Comment:
It is probably good idea to move this piece of code that joins together file filter and expanded locations together to the VirtualFolder model. That way all code that deals with location and filtering is kept in the model so it is easier to refactor/expand.